active/0000-ember-cleanup.md

```diff
@@ -0,0 +1,121 @@
+  - go through each line and check if it has one of `keys` from imports array, if yes delete found `key` from array, else do nothing
+  - when file is processed to the end check if initial imports array is empty, if yes do nothing, if not delete lines with imports that are not used
+
+---
+
+### controllers
+Deletes controllers with default content(just extending from `Ember.Controller`).
+
+Design:
+- get list of all files from `app/controllers/*`
+- iterate over each file and check if has only:
+ - `import Ember from 'ember'`
+ - `export default Controller = Ember.Controller.extend()`
+- if yes then delete file because there should be no difference between programatically auto-generated controller at runtime and this file
+
+---
+
+### templates
+Deletes templates with just `{{outlet}}` as their content.
+
+Design:
+- get list of files from `app/templates/*`
+- iterate over them
+- if file contains only `{{outlet}}` then delete it
+
+---
+
+### components
+Same pattern as `templates` could be followed for `components` except for `{{yield}}` instead of `{{outlet}}`.
+
+---
+
+### routes
+Same pattern as `controllers` could be followed for `routes` except for `Ember.Route.extend()` instead of `Ember.Controller.extend()`.
+
+---
+
+### unused
+Check for no longer referenced components, helpers, templates, controllers and eventually delete them.
+
+Design:
+##### helpers
+- get list of helpers from `app/helpers/*`
+- search for using helpers in templates as `{{helper}}` or `{{#helper}}`
+- if any occurence is found for concrete helper then don't delete it, else do it
+
+#### components
+Same pattern as with helpers but also check JavaScript files is any file imports this component(for example to extend from it).
+
+Design like:
+- Search for `/components/component-name` if anything is found then don't delete it component.
+- Else search for `./component-name` and check if file that has `./` import is in `app/components` directory, if yes, don't delete it, else do it.
+
+#### controllers
+- could be searched for `needs: ['controller-name']`
+- search for: `controller: 'controller-name'`
+- check if controller file path can be linked with any route from `router`
+- check if `controller` file is imported in any file
+- if no occurences found, delete file
+
+#### templates
+- check if it is used by any route, component
+- search for `templateName: 'template-name'`
+
+------------------
+At the end of each command summary should be presented to users with list of full paths of files that were deleted or affected by deleting imports.
+
+
+# Drawbacks
+
+I remember sometimes even empty controller file was needed for my application or default outlet file. I don't know if this is still an issue.
+
+# Alternatives
+
+Ember Doctor could be alternative but I think its purpose is to "heal" application not to "clean" it.
+
+# Unresolved questions
+
+- Could following `cleanups` be implemented?
```

Comment:
I believe the garbage collection heuristics are unfortunately far to ambiguous to accomplish anything but the most trivial cleanup. I believe this RFC actually belongs as a module of the proposed `ember doctor` command:

pr: https://github.com/ember-cli/ember-cli/pull/3193 (wip, could use some help if someone has time)
rfc: https://github.com/ember-cli/rfcs/blob/master/active/0001-ember-doctor.md